File: 0155-min-stack/0155-min-stack.py
"""
Time = O(1)
Space = O(n)
"""
class MinStack:

    def __init__(self):
        self.min_stack = []
        
    def push(self, val: int) -> None:
        if not self.min_stack:
            self.min_stack.append((val, val))
        else:
            min_ele = min(val, self.min_stack[-1][1])
            self.min_stack.append((val, min_ele))
        
    def pop(self) -> None:
        self.min_stack.pop()

    def top(self) -> int:
        return self.min_stack[-1][0]

    def getMin(self) -> int:
        return self.min_stack[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.cpp
#include <stack>

class MinStack {
public:
    MinStack() {
    }
    
    void push(int val) {
        pair<int, int> valPair;
        if (mStk.empty()) {
            valPair = make_pair(val, val);
        } else {
            valPair = make_pair(val, min(val, mStk.top().second));
        }
        mStk.push(valPair);
    }
    
    void pop() {
        mStk.pop();
    }
    
    int top() {
        return mStk.top().first;
    }
    
    int getMin() {
        return mStk.top().second;
    }

private:
    stack<pair<int, int>> mStk;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */